### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by pluskid** [pluskid/Mocha.jl#22]
Discussions and/or suggestions are welcome!
- Interface
  - Network architecture visualization
  - Recurrent Neural Networks
- Infrastructure
  - [ ] CUDA Stream
  - [ ] Multi-GPU support
  - [x] 4D tensor -> ND tensor (#21, #25)
  - [x] Unsupervised Learning ((deep) autoencoders and variants) (#29)
- Document
  - Developer's Guide

**Comment by the-moliver** [comment]
I'm planning to add time-delay neural networks. I have a working implementation ( https://github.com/the-moliver/NeuralNets.jl ) that I want to port to Mocha.

**Comment by philtomson** [comment]
It would be nice to have a Caffe file -> Mocha converter. Maybe I'll work on something like that. Should be doable, right? Or are there Caffe features that are not yet in Mocha?

**Comment by pluskid** [comment]
We already have the ability to load caffe models, but you still need to manually translate the model definition. Automatic translation of architecture is theoretically possible but I guess might by quite tedious to implement. (I'm thinking maybe there should be some universal Dnn architecture specification language coming out recently). Most of the core functionality in caffe has correspondence in mocha. But caffe also have many unofficial forks, which implemented some specific layers, for those, it is more difficult to convert.

**Comment by nikolaypavlov** [comment]
- It would be nice to have [maxout layer](http://jmlr.csail.mit.edu/proceedings/papers/v28/goodfellow13.pdf) in addition to dropout..
- a [Max-norm Regularization](http://www.cs.toronto.edu/~hinton/absps/JMLRdropout.pdf) would can help for optimal dropout nets tuning.

**Comment by pluskid** [comment]
@nikolaypavlov Thanks for the suggestions

* Based on my understanding, `maxout` is simply a max pooling over some units. We can achieve this by using the existing `PoolingLayer` or `ChannelPoolingLayer`. Let me know if you are talking about something else.
* Max-norm regularization is actually implemented, see for example [`filter_cons`](http://mochajl.readthedocs.org/en/latest/user-guide/layers/computation-layer.html#ConvolutionLayer.filter_cons) for `ConvolutionLayer`.

**Comment by nikolaypavlov** [comment]
@pluskid Great, I'll try to play with PoolingLayer.

**Comment by outlace** [comment]
Is this project meant to be the Theano/Torch of Julia?

Is there ever going to be OpenCL support?

**Comment by pluskid** [comment]
@outlace, this is more like torch than theano in that sense. There is no planned Opencl support unless Julia gets better native support for gpu targets.

**Comment by nstiurca** [comment]
I would be very interested in OpenCL support as well. In fact, I have half a mind to take a stab at it myself. If I can leverage an OpenCL BLAS library (say, [CLBLAS.jl](https://github.com/JuliaGPU/CLBLAS.jl)), then I basically just have to write `im2col.cl` and a couple of pooling and neuron kernels.

If I did this, in the interest of clarity would you be OK with renaming GPUBackend -> CUDABackend (adding `@deprecated typealias GPUBackend CUDABackend` or similar for compatibility), and naming the new backend `OCLBackend`?

**Comment by pluskid** [comment]
@nstiurca Thanks! This could be cool! Yes, I'm OK with the renaming if we have a working OpenCL backend!

**Comment by nstiurca** [comment]
OK, I will get started this weekend. Should we open an issue for the sake of tracking? Development-wise, it will be simplest for me to create an `opencl` branch on the fork of your project that  I already have. Do you prefer to have such a branch in your repo as well until OpenCL support is stable (assuming we get there...)? It might be good to do that for the sake of anyone else that wants to help develop OpenCL support.

**Comment by pluskid** [comment]
I would suggest do it in your branch, but open a pull request to here, with "[WIP]" in the title and description of the goal and current progress in the text (that you could updates periodically). I will not merge the pr until you have something reasonablely stable, but people will see the pr and could probably jump in to help.

**Comment by nstiurca** [comment]
That works for me. Look for it later today.

**Comment by outlace** [comment]
I think this is great. I currently have to use Torch because it's the only mature package that has an OpenCL backend. Being able to run models on my Macbook is fantastic. Really looking forward to this getting OpenCL support.

**Comment by nstiurca** [comment]
@outlace Caffe also has a fork with OpenCL support, but unfortunately for me I haven' been able to get either Torch nor Caffe to work on 32-bit ARM processor even though it has a fully compliant OpenCL 1.1.

Thus, I am going to start on rolling my own. See [PR #155](https://github.com/pluskid/Mocha.jl/pull/155).

**Comment by lqh20** [comment]
Any plans to implement batch normalization (http://jmlr.org/proceedings/papers/v37/ioffe15.pdf )? Looks like it's a great step forward in terms of trainging time!

**Comment by pluskid** [comment]
@lqh20 I'm recently joining a new project MXNet. We are building a julia interface called [MXNet.jl](https://github.com/dmlc/MXNet.jl). It is still at relatively early stage, but some features are already working. For example, batch normalization and multi-GPU training in the [cifar-10 example](https://github.com/dmlc/MXNet.jl/blob/master/examples/cifar10/cifar10.jl#L9) is already working quite nicely.

**Comment by philtomson** [comment]
Is MXNet.jl complementary to Mocha.jl or meant to replace it?

**Comment by pluskid** [comment]
@philtomson It depends. Mocha.jl still has its advantage of simplicity and portability. But in terms of computational efficiency or feature richness, I think MXNet.jl should be replacing Mocha.jl. Because it is built on top of libmxnet which is a language agnostic general deep learning library that is designed to have, for example, multi-GPU support. Moreover, the core component of libmxnet is being actively developed by a team, so in terms of features it is much better than Mocha.jl which is currently primarily developed by me in my very little free time. libmxnet itself is actually a joint efforts of authors from several different deep learning libraries.

**Comment by philtomson** [comment]
I wonder if mxnet could be an alternate backend for Mocha.jl? It seems like
that would preserve the advantages of Mocha.jl - simplicity, portability,
good documentation - while also allowing users to drop directly to your
MXNet.jl bindings if needed.

**Comment by pluskid** [comment]
@philtomson That could be one possible option. I will wait and see if that is feasible. As using MXNet.jl introduce an external dependency on libmxnet. If that dependency itself is not a problem, then using MXNet.jl directly might be a more viable option. Though a something still needs to be improved, esp. documents.

**Comment by philtomson** [comment]
On Sun, Oct 25, 2015 at 8:32 PM, Chiyuan Zhang <notifications@github.com>
wrote:

> @philtomson <https://github.com/philtomson> That could be one possible
> option. I will wait and see if that is feasible. As using MXNet.jl
> introduce an external dependency on libmxnet. If that dependency itself is
> not a problem, then using MXNet.jl directly might be a more viable option.
>
Right. The scenario I was thinking of was being able to keep the kind of
simple, declarative style of Mocha.jl while also being able to take
advantage of the performance of MXNet.jl.  Sure libmxnet has an advantage
of being "language independent", however, that can also be a weakness. It
could mean that you can't readily take advantage of powerful language
features specific to Julia, like macros, for example (or at least it might
be more difficult to do so). I suspect there's a lot of boilerplate code
required when you use libmxnet that could be eliminated at a higher level
of abstraction.

BTW: is the GPU backend of Mocha.jl not as performant as libmxnet? I guess
from what I understand of the docs mxnet allows for multiple GPUs whereas
Mocha.jl only allows for using one? If you compare performance between
Mocha.jl with only one GPU and libmxnet with only one GPU are they pretty
close?

I can see where training with multiple GPUs can be an advantage, but some
users might be running pre-trained models on a laptop with only a single
GPU (or some of us don't even have that as we only have an Intel integrated
GPU which doesn't do CUDA) and the current setup of Mocha.jl is actually
quite sufficient for doing this (people with this kind of setup wouldn't
notice any appreciable difference from using libmxnet, perhaps)

Also: Does the mxnet project have any plans for supporting OpenCL?


> Though a something still needs to be improved, esp. documents.
>

Mocha.jl's documents are actually pretty good at this point so this is a
problem for someone who tries moving from Mocha.jl to MXNet.jl. Using
Mocha.jl as a sort of a wrapper around MXNet.jl would mean you could
probably keep most of the documentation as is.

I suppose another idea would be to translate the CPP backend for Mocha.jl
to produce c++ code that make calls directly to libmxnet (or at least
paramatize it so that you could use openMP (as now) or libmxnet in the CPP
backend.

**Comment by philtomson** [comment]
I just got around to installing MXNet.jl and playing with it some. So far
it doesn't seem too much more difficult to use than Mocha,

**Comment by pluskid** [comment]
@philtomson Glad to hear that it works out nicely for you.

The single-GPU performance of Mocha.jl might be similar to MXNet.jl. MXNet.jl has a more flexible symbolic API to define network architectures, but internally optimizations are used to avoid unnecessary memory allocation & computation, etc. But multi-GPU is definitely a win on MXNet.jl side.

I agree that many users with small scale applications do not use GPUs. In this case, the default CPU only libmxnet.so should still be quite straightforward to compile (at least on Linux and OS X). And since libmxnet is actually relatively low level backend, many of the logics will still be built in Julia, and the interface is actually flexible and convenient enough to use.

One of the main goal of the joint-force under the dmlc/libmxnet is to avoid duplicated labors especially in the computational heavy backend. One layer implemented will be automatically available in Python, Julia, R frontends.

Currently I will be maintaining both Mocha.jl and MXNet.jl. In the future when MXNet.jl become more mature, I will try to advocate MXNet.jl as a successor of Mocha.jl.

**Comment by pluskid** [comment]
For those who is interested in RNN/LSTM in Julia. Here is [an char-rnn LSTM implementation in MXNet.jl](https://github.com/dmlc/MXNet.jl/tree/master/examples/char-lstm) now. It used explicit unrolling so everything fit in the current `FeedForward` model. For more general purpose variable length RNN without unrolling, we will still need to develop the modeling interface. I will add tutorial document soon.
